### PR TITLE
 make chat.el work even after user sends nothing 

### PR DESCRIPTION
**Pull request by lambdaloop**
This is an amazing package!

I was having some issues with it where I would sometimes press enter without typing anything and it would then fail in that session after that. This pull request fixes that bug.

### CODE REVIEW

File: chat.el [hunk 1]
```diff
@@ -127,12 +127,21 @@ If no key is found, error."
...

 
-(defun chat--async-raw (messages callback finalize)
+
+(defun remove-bad-messages (messages)
+  "Filters out items in MESSAGES with no content.
+MESSAGES is a list of items, where each item is a list of (key . value) pairs.
+Returns a new list of items where each item has a non-nil 'content' key.
+Makes it still work when user sends nothing in chat."
+  (seq-filter (lambda (item) (not (null (cdr (assoc "content" item))))) messages))
```
Comment (iwahbe) [inline chat.el]: You don't actually need `(not (null v))`, `v` is sufficient. `not` is actually an alias for `null`, so this is equivalent to `(not (not v))`.

```suggestion
  (seq-filter (lambda (item) (cdr (assoc "content" item))) messages))
```

**Review by iwahbe — CHANGES_REQUESTED** [review]
Hi @lambdaloop. I'm really glad you're enjoying `chat.el`. I can see and reproduce the bug. I think it would be cleaner if we fixed this in `chat--send-entries`:

```patch
 chat.el | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/chat.el b/chat.el
index 8e565df..26967b5 100644
--- a/chat.el
+++ b/chat.el
@@ -554,7 +554,7 @@ FINISH is called after all text has been inserted."
                               ('user "user")
                               ('bot "assistant")))
                  ("content" . ,(caddr entry))))
-             (reverse (cdr chat--entries)))
+             (reverse (cdr (seq-filter #'caddr chat--entries))))
      (lambda (chunk)
        (when chunk
          (with-current-buffer b
```

Do you want to update your PR, or should I do it?

**Comment by lambdaloop** [comment]
Oh that's really clever! I can change that now

**Comment by lambdaloop** [comment]
It is updated!

**Comment by lambdaloop** [comment]
Actually sorry I tested this further and it doesn't work like this.  It ignores the first message and then is always one message behind.  It seems that there is a (bot) entry at the beginning of the list which this ends up ignoring and breaking everything else.

Example of a set of `chat--entries`
```((bot 98) (user 73 "say something evil" 91) (bot 32 "Hello! How may I assist you today?" 66) (user 7 "say something nice" 25))```

**Comment by lambdaloop** [comment]
I've put the filtering back into `chat--async-raw` to make it work, simplifying the code based on your recommendations.

File: chat.el (hunk 1 above)
Comment (lambdaloop) [reply chat.el]: i ended up using this but just removed the function itself

**Comment by iwahbe** [comment]
> Actually sorry I tested this further and it doesn't work like this. It ignores the first message and then is always one message behind. It seems that there is a (bot) entry at the beginning of the list which this ends up ignoring and breaking everything else.
> 
> Example of a set of `chat--entries` `((bot 98) (user 73 "say something evil" 91) (bot 32 "Hello! How may I assist you today?" 66) (user 7 "say something nice" 25))`

You're exactly right! It's almost if there is a `cdr` that previously was used to clip off the `nil` valued `Bot >` message that the bot would type into, but is now clipping off the last thing the user said:
```emacs-lisp
             (reverse (cdr (seq-filter #'caddr chat--entries))))
                      ^^^^
```

Sorry about that. I believe that this fix works. Can you confirm that this also works for you?

```patch
 chat.el | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/chat.el b/chat.el
index 8e565df..6e8d375 100644
--- a/chat.el
+++ b/chat.el
@@ -554,7 +554,7 @@ FINISH is called after all text has been inserted."
                               ('user "user")
                               ('bot "assistant")))
                  ("content" . ,(caddr entry))))
-             (reverse (cdr chat--entries)))
+             (reverse (seq-filter #'caddr chat--entries)))
      (lambda (chunk)
        (when chunk
          (with-current-buffer b
```

<details>
<summary>
I've tested it and it works for me.
</summary>

```
You > 
You > 
You > So bye-bye miss...
Bot > American Pie, 
Drove my Chevy to the levee but the levee was dry, 
And them good old boys were drinking whiskey and rye, 
Singin' this'll be the day that I die.
You > 
You > 
You > 
You > So I called up the Captain
Bot > But the Captain he just sighed,
And said "I'm sorry to inform you son, but your Chevy's lost its drive"
And as I walked back to the crowd,
The flames had reached the sky,
And they played "Nearer My God to Thee"
The day the music died.
You > And clenching your fist for the ones like us
Bot > Who are oppressed by the figures of beauty,
You fixed yourself, you said, "Well, never mind,
We are ugly, but we have the music."

And then you got away, didn't you babe?
You just turned your back on the crowd.
You got away, I never once heard you say,
I need you, I don't need you,
I need you, I don't need you.
You > 
You > 
You > It looks like this works
Bot > Great! Is there anything else I can assist you with?
You > 
```

</details>

**Comment by lambdaloop** [comment]
Ahh I kept looking for the cdr somewhere else but I missed it right there, my bad! 
I've pushed a commit implementing your suggested fix. I just kept the cdr in there.

**Review by iwahbe — APPROVED** [review]
Thanks for your PR! LGTM!

**Pull request merged**